File: src/credential.ts
/**
 * Credential schema definition and validation for v1 credentials.
 *
 * Validates structure, types, whitespace, extra keys, and date correctness
 * without relying on the Date constructor (which silently rolls invalid dates).
 */

import { isValidDate } from './validation.js';

export type CredentialV1 = {
  authority: string;
  date: string;
  detail: string;
  honor: string;
  recipient: string;
  version: 1;
};

export type Credential = CredentialV1;

/** Pattern matching C0/C1 control characters and bidi overrides. Exported as a string constant (not a RegExp with `g`) to avoid `lastIndex` statefulness. */
export const CONTROL_CHAR_PATTERN = '[\\x00-\\x1f\\x7f-\\x9f\\u061c\\u200e\\u200f\\u202a-\\u202e\\u2066-\\u2069]';

/** Cached test-only regex (no `g` flag, safe to reuse — no `lastIndex` state). */
const CONTROL_CHAR_RE = new RegExp(CONTROL_CHAR_PATTERN);

/**
 * Strips C0/C1 control characters and bidi overrides. Does not handle
 * zero-width joiners or length limits — callers are responsible for truncation.
 *
 * Note: creates a new RegExp per call because the `g` flag carries `lastIndex`
 * state that would make a shared instance unsafe across calls.
 */
export function sanitizeForError(s: string): string {
  return s.replace(new RegExp(CONTROL_CHAR_PATTERN, 'g'), '');
}

export class UnsupportedVersionError extends Error {
  constructor(version: unknown) {
    const safeVersion = typeof version === 'number' ? String(version) : sanitizeForError(String(version));
    super(`Unsupported credential version: ${safeVersion}`);
    this.name = 'UnsupportedVersionError';
  }
}

const STRING_FIELDS = ['authority', 'date', 'detail', 'honor', 'recipient'] as const;
const ALL_FIELDS = new Set<string>([...STRING_FIELDS, 'version']);

const FIELD_MAX_LENGTHS = {
  authority: 200,
  recipient: 500,
  honor: 200,
  detail: 2000,
  date: 10,
} as const satisfies Record<(typeof STRING_FIELDS)[number], number>;

/** Validate an unknown value as a v1 Credential, throwing on any violation. */
export function validateCredential(obj: unknown): Credential {
  if (obj === null || typeof obj !== 'object' || Array.isArray(obj)) {
    throw new Error('credential must be a plain object');
  }

  const record = obj as Record<string, unknown>;

  // Version check first.
  if (!('version' in record)) {
    throw new Error('missing required field: version');
  }
  if (typeof record.version !== 'number') {
    throw new Error('version must be a number');
  }
  if (record.version !== 1) {
    throw new UnsupportedVersionError(record.version);
  }

  // Required string fields.
  for (const field of STRING_FIELDS) {
    if (!(field in record)) {
      throw new Error(`missing required field: ${field}`);
    }
    const value = record[field];
    if (typeof value !== 'string') {
      throw new Error(`${field} must be a string`);
    }
    if (value !== value.trim()) {
      throw new Error(`${field} must not have leading or trailing whitespace`);
    }
    if (CONTROL_CHAR_RE.test(value)) {
      throw new Error(`${field} contains invalid control characters`);
    }
    const codePointLength = [...value].length;
    if (codePointLength === 0) {
      throw new Error(`${field} must not be empty`);
    }
    if (codePointLength > FIELD_MAX_LENGTHS[field]) {
      throw new Error(`${field} exceeds maximum length of ${FIELD_MAX_LENGTHS[field]}`);
    }
  }

  // Reject extra keys.
  for (const key of Object.keys(record)) {
    if (!ALL_FIELDS.has(key)) {
      throw new Error(`unexpected field: ${key}`);
    }
  }

  // Date validation.
  if (!isValidDate(record.date as string)) {
    throw new Error(`invalid date: ${sanitizeForError(record.date as string)}`);
  }

  return obj as Credential;
}

File: go/core/credential.go
package core

import (
	"fmt"
	"regexp"
	"strings"
	"unicode/utf8"
)

// CredentialV1 represents a validated v1 credential.
type CredentialV1 struct {
	Authority string `json:"authority"`
	Date      string `json:"date"`
	Detail    string `json:"detail"`
	Honor     string `json:"honor"`
	Recipient string `json:"recipient"`
	Version   int    `json:"version"`
}

// controlCharPattern matches C0/C1 control characters and bidi overrides.
const controlCharPattern = `[\x00-\x1f\x7f-\x9f\x{061c}\x{200e}\x{200f}\x{202a}-\x{202e}\x{2066}-\x{2069}]`

var controlCharRE = regexp.MustCompile(controlCharPattern)

// SanitizeForError strips all control characters and bidi overrides from s.
func SanitizeForError(s string) string {
	return controlCharRE.ReplaceAllString(s, "")
}

// stringFields defines the required string fields in validation order.
var stringFields = []string{"authority", "date", "detail", "honor", "recipient"}

// allFields is the complete set of allowed fields.
var allFields = map[string]bool{
	"authority": true,
	"date":      true,
	"detail":    true,
	"honor":     true,
	"recipient": true,
	"version":   true,
}

// fieldMaxLengths defines the maximum rune count for each string field.
var fieldMaxLengths = map[string]int{
	"authority": 200,
	"date":      10,
	"detail":    2000,
	"honor":     200,
	"recipient": 500,
}

// ValidateCredential validates a parsed JSON object as a v1 credential.
// Input is map[string]any (from json.Unmarshal into an interface{}).
// Returns the validated credential or an error.
func ValidateCredential(obj map[string]any) (CredentialV1, error) {
	if obj == nil {
		return CredentialV1{}, fmt.Errorf("credential must be a plain object")
	}

	// 1. Version check.
	verRaw, ok := obj["version"]
	if !ok {
		return CredentialV1{}, fmt.Errorf("missing required field: version")
	}
	verNum, ok := verRaw.(float64)
	if !ok {
		return CredentialV1{}, fmt.Errorf("version must be a number")
	}
	if verNum != 1 {
		return CredentialV1{}, fmt.Errorf("Unsupported credential version: %v", verNum)
	}

	// 2. String fields — validated in order.
	values := make(map[string]string, len(stringFields))
	for _, field := range stringFields {
		raw, exists := obj[field]
		if !exists {
			return CredentialV1{}, fmt.Errorf("missing required field: %s", field)
		}
		s, ok := raw.(string)
		if !ok {
			return CredentialV1{}, fmt.Errorf("%s must be a string", field)
		}
		if s != strings.TrimSpace(s) {
			return CredentialV1{}, fmt.Errorf("%s must not have leading or trailing whitespace", field)
		}
		if controlCharRE.MatchString(s) {
			return CredentialV1{}, fmt.Errorf("%s contains invalid control characters", field)
		}
		if s == "" {
			return CredentialV1{}, fmt.Errorf("%s must not be empty", field)
		}
		if utf8.RuneCountInString(s) > fieldMaxLengths[field] {
			return CredentialV1{}, fmt.Errorf("%s exceeds maximum length of %d", field, fieldMaxLengths[field])
		}
		values[field] = s
	}

	// 3. Extra keys.
	for key := range obj {
		if !allFields[key] {
			return CredentialV1{}, fmt.Errorf("unexpected field: %s", key)
		}
	}

	// 4. Date validation.
	if !IsValidDate(values["date"]) {
		return CredentialV1{}, fmt.Errorf("invalid date: %s", SanitizeForError(values["date"]))
	}

	return CredentialV1{
		Authority: values["authority"],
		Date:      values["date"],
		Detail:    values["detail"],
		Honor:     values["honor"],
		Recipient: values["recipient"],
		Version:   1,
	}, nil
}
